lib/vector-core/src/fanout.rs

```diff
@@ -42,13 +37,66 @@ impl Fanout {
 
         let fanout = Self {
             sinks: vec![],
-            i: 0,
             control_channel: control_rx,
         };
 
         (fanout, control_tx)
     }
 
+    // TODO make a real error
+    pub async fn send_all(&mut self, events: Vec<Event>) -> Result<(), ()> {
+        self.process_control_messages().await;
+
+        if self.sinks.is_empty() {
+            // nothing, intentionally
+        } else if self.sinks.len() == 1 {
+            if let Some(sink) = &mut self.sinks[0].1 {
+                for event in events {
+                    sink.feed(event).await?;
+                }
+                sink.flush().await?;
+            }
+        } else {
+            let mut faulty_sinks = Vec::with_capacity(0);
+
+            {
+                let mut jobs = FuturesUnordered::new();
+                let mut clone_army: Vec<Vec<Event>> = Vec::with_capacity(self.sinks.len());
+                for _ in 0..(self.sinks.len() - 1) {
+                    clone_army.push(events.clone());
+                }
+                clone_army.push(events);
+
+                for (id, ms) in self.sinks.iter_mut() {
+                    if let Some(sink) = ms.as_mut() {
+                        let events: Vec<Event> = clone_army.pop().unwrap();
+                        jobs.push(async move {
+                            for event in events {
+                                sink.feed(event).await.map_err(|e| (id.clone(), e))?
+                            }
+                            sink.flush().await.map_err(|e| (id.clone(), e))
+                        });
+                    }
+                }
+
+                while let Some(res) = jobs.next().await {
```

Comment:
I'm wondering if we need to `select` over this and the control channel here. For example, if a sink is blocked and there's a config reload to remove it, we'd want to drop the sink asap rather than wait (potentially forever) for a successful `send` here.

Reply:
That is a good point, actually.

Reply:
Oh wait, actually, I think this is okay? The reload happens while we're here in this loop. The sink feed/flush fails, that error bubbles up here, the sink is marked as faulty and nuked. At least, my understanding was that when a shutdown of a component happened write/flushes into it would fail appropriately. Is that incorrect?

Reply:
IIRC, normal shutdowns essentially involve disconnecting the component's input and letting the stream it's reading end naturally. The only place that we really thread a shutdown signal is into sources. So from `Fanout`'s perspective, it's going to see a `Remove` message come through before the component itself is actually stopped.

Reply:
Interesting, hmm. So, by the time we hit `send_all` part of that stream has been read -- whatever was ready to go up to a certain size -- and delivered here. We hit this loop, the caller has stopped reading from the stream by now and the sink is still alive to process the future against it. Next call to `send_all` the control messages are polled, stopped sink is removed. 

I think that logic is accurate and we don't need to poll the control messages here?

Reply:
> the caller has stopped reading from the stream by now

This part I think is slightly off. The caller of `send_all` is the source or transform producing the event stream, which will continue independent of one of its consumers being removed. So the situation I'm talking about is something like the following:

1. Source A running, and its output is being fed to Sinks B and C. So the caller of `send_all` is Source A's pump task and the `Fanout`'s `sinks` field consists of the buffer senders for Sinks B and C.
2. Sink C hits some condition where it is not making progress, leading us to fill its buffer.
3. As Source A continues to produce data, we push a job for Sink C into the `FuturesUnordered` that becomes blocked on `sink.feed(event).await`.
4. The whole `Fanout` (and eventually Source A itself) becomes blocked on `jobs.next().await` waiting for Sink C to make progress (this is the normal, expected behavior).
5. A config reload happens that intends to remove Sink C from the config.
6. The control message is never read because `Fanout` remains blocked, meaning that the config reload itself is stalled with the blocked sink that being removed. To recover, the user would need to restart the whole Vector process.

If we instead do a select between `jobs.next()` and `self.controll_channel.next()`, we'd have the opportunity to see that we're dropping Sink C and stop waiting for the send to succeed, allowing the config reload to complete and the whole system to start making progress again.

Reply:
Ah, yep. I see it.

Reply:
Hmm. Does that break this whole patch? So, right now we can't associate a Future stuffed into FuturesUnordered with a name, which is what the `ControlMessage` communicates. In order for selecting on the control channel to work we have to be able to manipulate the vector of sinks and ignore / cancel any future associated with the specific `GenericEventSink`. 
In your example above, same scenario, but by the time point 6 comes around we poll the control channel and know which named sink is removed but can't get at the future to cancel it. 

I suppose we could, instead of using a Future, use a JoinHandle instead and implement effectively the upcoming tokio [`JoinSet`](https://github.com/tokio-rs/tokio/pull/4335/) but with reference by name capability too but that seems like a non-trivial leap in complexity here.

Reply:
I was just trying to think through the same thing. It feels like there should be some way to make the send task cancellable (joining with a per-sink `tokio::sync::Notify` that we trigger when a sink is being removed, or similar), but it's definitely a significant bump in complexity.

Reply:
This makes me wonder, actually, if we aren't overloading what the `Fanout` does. So, I can imagine a `Fanout` whose sold job is to process incoming `Vec<Event>` instances and not control messages working just fine, relatively simply. We stick the processing of control messages one level up and create brand new `Fanout` instances on reload, newly wired: shut down the old `Fanout`, plug in the stream to the new `Fanout` and let it go.

Reply:
That is an interesting idea. I still think you have some form of the same issue, where you want to be able to abort writes to a specific sink when it is removed. If `Fanout` held a fixed set of sinks and was replaced wholesale on reload, you wouldn't really have a way to know what it was blocking on when you went to replace it. You could have it expose that data somehow, but that feels like it starts to lose the nice layering and simplicity.

Reply:
True. What I was thinking was when the `Fanout` flips over into shutdown mode it is responsible for attempting to poll every outstanding `Future`, cancel those that aren't ready. But now that I think about it that's dicey because any `Event` we've pushed in has the potential to be lost and I can think of a scenario where reloads happen rapidly enough to mean there's no actual `Event` transmission.